imports/plugins/core/catalog/client/components/publishControls.js

```diff
@@ -42,15 +65,28 @@ class PublishControls extends Component {
     return (
       <PrimaryAppBar>
         {this.renderChangesNotification()}
-        <Button
-          color="primary"
-          variant="contained"
-          disabled={Array.isArray(documentIds) && documentIds.length === 0}
-          label="Publish"
-          onClick={onPublishClick}
-        >
-          {i18next.t("productDetailEdit.publish")}
-        </Button>
+        <Mutation mutation={publishProductsToCatalog}>
+          {(mutationFunc, { data, error }) => {
+            if (error) {
+              Alerts.toast(error.message, "error");
+            }
+            if (data) {
+              Alerts.toast(i18next.t("admin.catalogProductPublishSuccess"), "success");
+            }
```

Comment:
It doesn't work to do the alerts this way because the component rerenders potentially unlimited times and every rerender pops up a new toast alert. You can see this when testing. After you change something and then click publish, every time you change another field or do anything, an additional "successfully published" alert appears.

You could move the alerts into the other functions, after the `await mutation` lines. Otherwise, `Mutation` takes `onCompleted` and `onError` callbacks, but I think I've had trouble getting those to work consistently in the past. They may have fixed since then. See https://www.apollographql.com/docs/react/v2.4/essentials/mutations/#props

You have similar code with a similar issue in the grid file, but in that case when I tested I actually never saw any toasts, even though both selected products were successfully published and their table cells updated to say "Published".

Reply:
Updated using callbacks, they seem to be working fine.